http example3: explicit mux provides real value

extemporalgenome opened the pull request:
- explicit mux use now provides value over default mux.
- users slice protected by RWMutex -- avoids previous race.
- full (but short) web 1.0 example; easily changed into SPA.

ardan-bkennedy commented on 12-http/example3/main.go:
```diff
@@ -18,20 +21,46 @@ type user struct {
+	// initialize users, so that the output JSON won't be null
+	// prior to adding users.
+	users = make([]user, 0)
+)
```
Let's not put these in a block.

extemporalgenome replied:
That's idiomatic Go style. _Always_ put a lock in a block ahead of the data it protects. If the lock is part of a struct type declaration, put the lock field immediately before the field(s) it protects, and if there are other fields not protected by the mutex, put a blank line between the protected fields and the other fields.

ardan-bkennedy replied:
Only if they are related and the var can be the comment for everything. These are not variables that could be grouped together under a single comment.

extemporalgenome replied:
There aren't a lot of stdlib examples of global mutexes, of course, therefore there won't be any specific stylistic convention of comments vs global var groups involving mutexes. Having the data and the mutex as globals is flakely and artificial anyway.

I'll rework it to use a type involving a mutex (for which there are numerous stylistic precedents I can follow), and define methods on that type to act as the handlers. It'll use about the same amount of code, but provide a good example of building your own `http.Handler`. It'll also reinforce method values that we saw in the methods lesson, since I'll be using some methods as `http.HandlerFunc` values.

ardan-bkennedy replied:
I think it may be best to push it and then I can do some styling changes and we can meet in the middle.

ardan-bkennedy commented on 12-http/example3/main.go:
```diff
@@ -18,20 +21,46 @@ type user struct {
+
+// index page template
+var IdxTpl = template.Must(template.ParseFiles("template/index.html"))
+
```
Does this need to be exported

extemporalgenome replied:
Nope. It's package main though, so except in rare circumstances, exporting rules are more gray. I'll unexport it though.

ardan-bkennedy commented on 12-http/example3/main.go:
```diff
@@ -39,7 +68,9 @@ func usersHandler(w http.ResponseWriter, r *http.Request) {
 	switch r.Method {
 	case "GET", "HEAD":
 		// List the users
+		mu.RLock()
```
Can we use the syntax I teach

RLock()
{
   code
}
RUnlock()

extemporalgenome replied:
Don't forget the old saying: "Lock data, not code"

I'll add the anonymous block, but we need to make it extra clear that critical sections are an unfortunate side-effect of how locks work -- they're not the purpose of locks. High level mutex semantics in Go, like in many other C-derived languages, are a mental construct only (on the other hand, channels have language-defined high level semantics, which makes them safer and more powerful when used correctly).

extemporalgenome replied:
http://www.makelinux.net/books/lkd2/ch09lev1sec2 explains some of the "don't lock code" reasoning in the "What do I lock?" section.

ardan-bkennedy commented on 12-http/example3/main.go:
```diff
@@ -18,20 +21,46 @@ type user struct {
 
-	// This will act as a catch all for the rest of the routes
-	mux.Handle("/", http.FileServer(http.Dir("public")))
+	http.Handle("/api/v1/", http.StripPrefix("/api/v1", api))
```
Please add a comment as to why to use Strip?

We don't need the FileServer anymore?

extemporalgenome replied:
fs could have been created on the fly each time, but if http.FileSystem allocates at all, this will be more efficient. Just like we could have parsed the index template every time we used it -- but that would've been inefficient.

ardan-bkennedy commented on 12-http/example3/main.go:
```diff
@@ -18,20 +21,46 @@ type user struct {
+		fs.ServeHTTP(w, r)
+		return
+	}
+	mu.RLock()
```
extra line of spacing

ardan-bkennedy commented on 12-http/example3/main.go:
```diff
@@ -18,20 +21,46 @@ type user struct {
+
+func baseHandler(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Path != "/" {
+		fs.ServeHTTP(w, r)
```
is this handling the file service?

extemporalgenome replied:
Yeah. Stylistically do the shorter-code thing first (in this case, falling back to static file serving if the path isn't exactly "/"). Since we're using an explicit Mux on /api/v1 in func main, then anything with an /api/v1 prefix won't get handled by baseHandler.

ardan-bkennedy commented on 12-http/example3/main.go:
```diff
@@ -18,20 +21,46 @@ type user struct {
+// index page template
+var IdxTpl = template.Must(template.ParseFiles("template/index.html"))
+
+// file server for any future assets and other static files
```
// fs represents a file server .....

extemporalgenome replied:
"represents" is generally a word used in type declaration doc comments. I'll use the word "is".

ardan-bkennedy commented on 12-http/example3/main.go:
```diff
@@ -18,20 +21,46 @@ type user struct {
+	users = make([]user, 0)
+)
+
+// index page template
```
// idxTpl .....

ardan-bkennedy commented on 12-http/example3/main.go:
```diff
@@ -18,20 +21,46 @@ type user struct {
+var (
+	// mu protects users.
+	mu sync.RWMutex
+	// initialize users, so that the output JSON won't be null
```
// users ....

extemporalgenome replied:
please clarify what you'd like to see here.

ardan-bkennedy commented on 12-http/example3/main.go:
```diff
@@ -18,20 +21,46 @@ type user struct {
+	log.Fatalln(http.ListenAndServe(bind, nil))
+}
+
+func baseHandler(w http.ResponseWriter, r *http.Request) {
```
function comments  baseHandler....

extemporalgenome replied:
Will do

extemporalgenome commented:
I've addressed all the comments that we both agreed on but which also did not prove unwieldy. For a number of reasons, we need to discuss the use of anonymous blocks to denote critical sections in this particular example.

The comments in the var decl group now exactly match the stdlib style for fields inside struct type declarations.

Feel free to merge it and make some stylistic tweaks. I'll create a separate branch/pull-request later on with the changes I discussed. We can evaluate those separately when that pull request is made.
